shared-bindings/ps2io/Ps2.c

```diff
@@ -0,0 +1,194 @@
+//|
+//| Ps2 implements the PS/2 keyboard/mouse serial protocol, used in
+//| legacy devices. It is similar to UART but there are only two 
+//| lines (Data and Clock). PS/2 devices are 5V, so bidirectional
+//| level converters must be used to connect the I/O lines to pins
+//| of  3.3V boards.
+//|
+//| .. class:: Ps2(data_pin, clock_pin)
+//|
+//|   Create a Ps2 object associated with the given pins.
+//|
+//|   :param ~microcontroller.Pin data_pin: Pin tied to data wire.
+//|   :param ~microcontroller.Pin clock_pin: Pin tied to clock wire.
+//|   This pin must spport interrupts.
+//|
+//|     import ps2io
+//|     import board
+//|
+//|     kbd = ps2io.Ps2(board.D10, board.D11)
+//|
+//|     while len(kbd) == 0:
+//|         pass
+//|     print(kbd.get_byte())
+//|     print(kbd.send_byte(0xed))
+//|     print(kbd.send_byte(0x02))
+//|
+STATIC mp_obj_t ps2io_ps2_make_new(const mp_obj_type_t *type, size_t n_args, const mp_obj_t *pos_args, mp_map_t *kw_args) {
+    enum { ARG_datapin, ARG_clkpin };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_datapin, MP_ARG_REQUIRED | MP_ARG_OBJ },
+        { MP_QSTR_clkpin, MP_ARG_REQUIRED | MP_ARG_OBJ },
+    };
+    mp_arg_val_t args[MP_ARRAY_SIZE(allowed_args)];
+    mp_arg_parse_all(n_args, pos_args, kw_args, MP_ARRAY_SIZE(allowed_args), allowed_args, args);
+    assert_pin(args[ARG_clkpin].u_obj, false);
+    assert_pin(args[ARG_datapin].u_obj, false);
+    const mcu_pin_obj_t* clkpin = MP_OBJ_TO_PTR(args[ARG_clkpin].u_obj);
+    assert_pin_free(clkpin);
+    const mcu_pin_obj_t* datapin = MP_OBJ_TO_PTR(args[ARG_datapin].u_obj);
+    assert_pin_free(datapin);
+
+    ps2io_ps2_obj_t *self = m_new_obj(ps2io_ps2_obj_t);
+    self->base.type = &ps2io_ps2_type;
+
+    common_hal_ps2io_ps2_construct(self, datapin, clkpin);
+
+    return MP_OBJ_FROM_PTR(self);
+}
+
+//|   .. method:: deinit()
+//|
+//|      Deinitialises the Ps2 and releases any hardware resources for reuse.
+//|
+STATIC mp_obj_t ps2io_ps2_deinit(mp_obj_t self_in) {
+    ps2io_ps2_obj_t *self = MP_OBJ_TO_PTR(self_in);
+    common_hal_ps2io_ps2_deinit(self);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_1(ps2io_ps2_deinit_obj, ps2io_ps2_deinit);
+
+//|   .. method:: __enter__()
+//|
+//|      No-op used by Context Managers.
+//|
+//  Provided by context manager helper.
+
+//|   .. method:: __exit__()
+//|
+//|      Automatically deinitializes the hardware when exiting a context. See
+//|      :ref:`lifetime-and-contextmanagers` for more info.
+//|
+STATIC mp_obj_t ps2io_ps2_obj___exit__(size_t n_args, const mp_obj_t *args) {
+    (void)n_args;
+    common_hal_ps2io_ps2_deinit(args[0]);
+    return mp_const_none;
+}
+STATIC MP_DEFINE_CONST_FUN_OBJ_VAR_BETWEEN(ps2io_ps2___exit___obj, 4, 4, ps2io_ps2_obj___exit__);
+
+//|   .. method:: get_byte()
+//|
+//|     Removes and returns the oldest received byte. When buffer
+//|     is empty, returns a negative number.
```

Comment:
Could this be popleft() to match PulseIn? Instead of returning a negative number on error please raise an Exception.

Reply:
Agreed.